### ISSUE TITLE
Floating point casts are incorrectly reported as redundant

### ISSUE THREAD

**Issue by SLaks** [dotnet/roslyn#6309]
http://blog.paranoidcoding.com/2014/12/22/redundant-cast.html

Example:

``` C#
float X() => 2 / (float)X();
```

Removing the `(float)` [changes the generated IL](http://tryroslyn.azurewebsites.net/#f:>ilr/K4Zwlgdg5gBAygTxAFwKYFsDcAoADsAIwBswBjGUogQxBBgGEYBvGV7V1gMyIHsrkYADQAUAShgBeAHwwATDAD0MYdz7JRI0TgC+QA==), yet the IDE still grays it.

**Comment by CyrusNajmabadi** [comment]
Jared interestingly mentioned: "This brings up another question: Is this behavior guaranteed by the C# specification? At this time the answer is no"

If it's not guaranteed behavior, then I think it's ok to consider the cast redundant.  Plus, I think it's good to change the behavior of non-guaranteed parts of the compiler with every release.  That way people don't take dependencies on things they shouldn't.

**Comment by SLaks** [comment]
Interesting point.
@jaredpar  [disagrees](https://twitter.com/jaredpar/status/658305503269777408), though.

**Comment by mburbea** [comment]
As far as I can tell the CLR considers the cast as non-redundant. I'd prefer if we get a modified so that I can guarantee strict fp math every where.
see #3896 for information, and the relevant part of the spec.

**Comment by jaredpar** [comment]
@CyrusNajmabadi that blog post is a little dated on that particular point.  This was eventually discussed amongst the compiler team and Mads and we decided to make it behavior guaranteed by the language spec.  Reason being that it's common practice in other languages and it's reasonable to believe programs out there depend on this behavior.

I have first hand experience with the latter point.  That blog post began because I removed that behavior from the compiler (no comments, thought it was a mistake) at one point and I broke parts of the research project I was involved in as a result.  Digging into the failure eventually lead me to writing that post and learning a bit about floating point.

**Comment by DustinCampbell** [comment]
So... is this a bug or not? :smile:

**Comment by CyrusNajmabadi** [comment]
Well, if the behavior is guaranteed by the spec, and this changes behavior... then it seems like a bug to me :)

How severe it is is a different matter.

**Comment by bgrainger** [comment]
> Mads and we decided to make it behavior guaranteed by the language spec

@jaredpar Is there an updated C# language spec that includes this guarantee? If so, which section is it in? (The latest spec I can find online is version 5.0, but of course this implementation detail isn't mentioned in sections 4.1.6 or 6.2.1.)

**Comment by jaredpar** [comment]
@bgrainger it won't appear until the C# 6.0 spec is released.  I'm not sure what the ETA is for that.

**Comment by DustinCampbell** [comment]
So, I started taking a look at this and the expected behavior is a bit unclear to me. Remove Redundant Cast does not aspire to be IL-preserving. After all, the IDE is not the compiler. So, I'm looking at this problem with an eye toward finding a solution that keeps the user's program correct while still providing value by removing casts where truly redundant, including scenarios where redundant IL is generated. Consider the following code:

```C#
class C
{
    float f;
    float X() => 42.0f;
    void M()
    {
        this.f = (float)X();
    }
}
```

The compiler will generate code like so:

```
.method private hidebysig instance void  M() cil managed
{
  // Code size       14 (0xe)
  .maxstack  8
  IL_0000:  ldarg.0
  IL_0001:  ldarg.0
  IL_0002:  call       instance float32 C::X()
  IL_0007:  conv.r4
  IL_0008:  stfld      float32 C::f
  IL_000d:  ret
} // end of method C::M
```

However, I'm not sure that the "conv.r4" is actually necessary in this case since the value on the top of the stack will immediately stored in a field of type float32. I suspect that the compiler is simply being paranoid and this instruction is truly necessary.

**Comment by DustinCampbell** [comment]
@jaredpar and I chatted about this and decided that the floating casts should be considered redundant in the following cases:

* A cast to a literal of the same type. (```(float)1.0f```)
* A cast to a cast of the same type. (```(float)(float)x```)
* A cast that is immediately assigned to a storage location, that is, a field, local or array element. (```this.myFloat = (float)X()```)

**Comment by bgrainger** [comment]
> A cast that is immediately assigned to a storage location, that is, a field, local or array element. 

According to CLI I.12.1.3, the CLI doesn't consider a "local" to be a "storage location" (if this statement is using the CLI definition of the term; emphasis added):

> Storage locations for floating-point numbers (**statics, array elements, and fields of classes**) are of fixed size. ... Everywhere else (on the evaluation stack, as arguments, as return types, and as **local variables**) floating-point numbers are represented using an internal floating-point type.  In each such instance, the nominal type of the variable or expression is either `float32` or `float64`, but its value can be represented internally with additional range and/or precision.

It even appears to say that floating-point values placed in a "storage location" (field, local, or array element) can continue to use the higher precision of the internal representation:

> When a floating-point value whose internal representation has greater range and/or precision than its nominal type is put in a storage location, it is automatically coerced to the type of the storage location. This can involve a loss of precision or the creation of an out-of-range value (NaN, +infinity, or -infinity). However, the value might be retained in the internal representation for future use, if it is reloaded from the storage location without having been modified. It is the responsibility of the compiler to ensure that the retained value is still valid at the time of a subsequent load, taking into account the effects of aliasing and other execution threads (see memory model (§I.12.6)). (This freedom to carry extra precision is not permitted, however, following the execution of an explicit conversion (`conv.r4` or `conv.r8`), at which time the internal representation must be exactly representable in the associated type. 

An additional remark makes it seem like the C# compiler should continue to emit `conv.r4` instructions when there is an explicit cast to float, even when storing in a local or field of `float32` type:

> At the same time, CIL generators can force operations to respect language-specific rules for representations through the use of conversion instructions. 

In summary, I think a cast to `float` before storing in a field or local may not be redundant from the CLI's perspective.

**Comment by bgrainger** [comment]
OTOH, if controlling this particular behaviour of the CLI is not a feature of the C# compiler (obviously, not everything expressible in IL has to be expressible in C#) then it's clear that the cast could be considered redundant (from the C# perspective).

**Comment by DustinCampbell** [comment]
@bgrainger -- bear in mind that we're talking about the IDE experience here and not the compiler. The IDE is permitted to take a more heuristic approach than the compiler.

According to the CLI, it seems that it is perfectly reasonable for the following two methods to return different results:

```C#
float M1(float x, float y)
{
    return (float)x / (float)y;
}

float M2(float x, float y)
{
    return x / y;
}
```

Is that correct? This seems very disturbing to me.

**Comment by SamB** [comment]
@DustinCampbell: Is that a regression compared to C and C++? (Maybe we need some kind of `-ffast-math` / `-fno-fast-math` like directive to the JIT, but how would you ensure that the "right" values applied to possibly-inlined math library functions?)

**Comment by jcouv** [comment]
@DustinCampbell FYI The same kind of logic is being added to VB as well (https://github.com/dotnet/roslyn/pull/22552). Some casts that seem redundant are in fact useful.
